rpcs3/Emu/RSX/Common/ProgramStateCache.h

```diff
@@ -214,16 +214,20 @@ class program_state_cache
 
 			lock.upgrade();
 			auto [it, inserted] = m_fragment_shader_cache.try_emplace(rsx_fp);
-			new_shader = &(it->second);
-			recompile = inserted;
+			new_shader          = &(it->second);
+			recompile           = inserted;
 
-			if (inserted)
+			if (recompile)
 			{
 				it->first.clone_data();
```

Comment:
Not necessary to clone under lock. The fix for this is literally just move the bracket on line 225 (prior) to before this if (inserted) block and you're done. Its a two line change.

Reply:
You'll need to declare it outside the block for this to work. clone_data does a bunch of memcpy operations.

Reply:
Alrighty, C++ is not my forte, but I think I figured out the right way to unpack the tuple. Let me know if there's a better way to do it. :)

Reply:
lgtm